components/places/ios/Places/Places.swift

```diff
@@ -0,0 +1,646 @@
+            })
+        }
+    }
+
+    /**
+     * Returns the bookmark subtree rooted at `rootGUID`.
+     *
+     * This differs from `getBookmark` in that it populates folder children
+     * recursively (specifically, any `BookmarkFolder`s in the returned value
+     * will have their `children` list populated, and not just `childGUIDs`.
+     *
+     * However, if `recursive: false` is passed, only a single level of child
+     * nodes are returned for folders.
+     *
+     * - Parameter rootGUID: the GUID where to start the tree.
+     *
+     * - Parameter recursive: Whether or not to return more than a single
+     *                        level of children for folders. If false, then
+     *                        any folders which are children of the requested
+     *                        node will *only* have their `childGUIDs`
+     *                        populated, and *not* their `children`.
+     *
+     * - Returns: The bookmarks tree starting from `rootGUID`, or null if the
+     *            provided guid didn't refer to a known bookmark item.
+     * - Throws:
+     *     - `PlacesError.databaseCorrupt`: If corruption is encountered when fetching
+     *                                      the tree
+     *     - `PlacesError.databaseInterrupted`: If a call is made to `interrupt()` on this
+     *                                          object from another thread.
+     *     - `PlacesError.connUseAfterAPIClosed`: If the PlacesAPI that returned this connection
+     *                                            object has been closed. This indicates API
+     *                                            misuse.
+     *     - `PlacesError.databaseBusy`: If this query times out with a SQLITE_BUSY error.
+     *     - `PlacesError.unexpected`: When an error that has not specifically been exposed
+     *                                 to Swift is encountered (for example IO errors from
+     *                                 the database code, etc).
+     *     - `PlacesError.panic`: If the rust code panics while completing this
+     *                            operation. (If this occurs, please let us know).
+     */
+    func getBookmarksTree(rootGUID: String, recursive: Bool) throws -> BookmarkNode? {
+        return try queue.sync {
+            try self.checkApi()
+            let buffer = try PlacesError.unwrap { (error: UnsafeMutablePointer<PlacesRustError>) -> PlacesRustBuffer in
+                if recursive {
+                    return bookmarks_get_tree(self.handle, rootGUID, error)
+                } else {
+                    return bookmarks_get_by_guid(self.handle, rootGUID, 1, error)
+                }
+            }
+            if buffer.data == nil {
+                return nil
+            }
+            defer { places_destroy_bytebuffer(buffer) }
+            // This should never fail, since we encoded it on the other side with Rust,
+            // should we use `try! instead?
```

Comment:
Follow-up? 😄

Reply:
The iOS team have always asked us to change try! to try wherever possible, so it's probably fine. I should maybe delete asking if we should use try!